### DISCUSSION TITLE
Stochastic gradients in pytensor

### DISCUSSION THREAD

**Discussion by jessegrabowski** [pymc-devs/pytensor#1424] [Q&A]
### Description

There are many problems in machine learning that require differentiating through random variables. This specifically came up in the context of https://github.com/pymc-devs/pymc/pull/7799, but it was also implicated in #555. 

Right now, `pt.grad` refuses to go through a random variable. That's probably the correct behavior. But we *could* have another function, `pt.stochastic_grad`, that would do it and return a stochastic gradient. It would also take an `n_samples` argument, since what we're actually computing is a sample-based MCMC estimate of the gradients with respect to parameters.

Most often, the RV would have to know a "reparameterization trick" to split its parameters from the source of randomness. The canonical example is the non-centered normal parameterization. Given a loss function $\mathcal L$ that depends on $x \sim N(\mu, \sigma)$ , the proposed `pt.stochastic_grad` would compute the gradient of the expected loss given the RVs: $\nabla_\theta \mathbb{E}_x [\mathcal{L(g(x, \theta)})] = \mathbb{E} \nabla_\theta \mathcal{L}(g(x, \theta))$. The so-called reparameterization trick just does a non-centered parameterization, $x = \mu + \sigma z, \quad z \sim N(0,1)$, so that now the gradient contribution of $g(x, \theta)$ can be estimated:

$$
\approx \frac{1}{N} \sum_{i=1}^N \nabla_\theta L(\mu + \sigma z_i)
$$

And the (expected) sensitivity equations for the parameters of the normal are:

$$
\begin{align}
\bar{\mu} &= \frac{1}{N} \sum_{i=1}^N \frac{\partial L}{\partial x^{(i)}}, \\ 
\bar{\sigma} &= \frac{1}{N} \sum_{i=1}^N \frac{\partial L}{\partial x^{(i)}} \cdot z^{(i)}
\end{align}
$$

It would be easy enough for a normal_rv to know this, and to supply these formulas when requested to by the hypothetical `pt.stochastic_grad`.

I guess other RVs also have reparameterizations (beta, dirichlet, gamma, ...?), but in some cases, there are multiple options but it's not clear which one is best to use in what cases. Some thought would have to be given to how to handle that.

When a reparameterization doesn't exist, there are other, higher-variance options to compute the expected gradients (the REINFORCE gradients, for example). We could offer these as a fallback.

Basically, this issue is proposing this API, and inviting some discussion on whether we want this type of feature, and how to do it if so. The `pt.stochastic_grad` function would be novel as far as I know. Other packages require that you explicitly generate samples in your computation graph. For example, `torch` offers `normal(mu, sigma).rsample(n_draws)`, which generates samples using reparameterization trick, so the standard `loss.backward()` works. Here the user can't "accidentally" trigger stochastic gradients (because you have to call rsample instead of sample).

I'm less familiar with how numpyro works, but I believe that something like `numpyro.sample("z", dist.Normal(mu, sigma))` automatically implies reparameterization trick if it's available. They don't have a special idiom like `rsample` for when it will or won't be used.

**Comment by ricardoV94** [comment]
CC @aseyboldt

**Comment by jessegrabowski** [comment]
Also @zaxtax had some thoughts on this during the VI hack

**Comment by lucianopaz** [comment]
I'm bringing this up again. I'm totally in favor of having the extra `pt.stochastic_grad` function. The stochastic gradient is relevant for other problems besides VI. For example, it comes up in the methods for prior elicitation based on prior predictive simulations:
- https://proceedings.mlr.press/v124/hartmann20a/hartmann20a.pdf
- https://www.nature.com/articles/s41598-024-68090-7.pdf
- https://arxiv.org/pdf/2411.15826

From the implementation side of things, it looks like `pt.grad` relies on the `Op.L_op` to get the gradient along with the chain rule. We could add a new `stochastic_L_op` method to our `Op` that calls `L_op` by default, but overload this for random variables. We could also reuse the code from `grad` by adding some flag to tell it to use `stochastic_L_op` instead of `L_op`. It would then be up to the random variable Op's to define a reparametrized version from which to compute the `stochastic_L_op`.

**Comment by ricardoV94** [comment]
The other options considered where to have a dataset of "parametrization_tricks" rewrites that gets called inside `schochastic_grad/jacobian` before attempting to take regular grad.

I also checked what JAX does, and it's seem basically the case that all they're RVs are implemented as `f(rng.random(), *params)`, so it's always in the "reparametrized" form from the get go.  We do want to do something like that for some variables, as it allows optimizing more computation that's now hidden in the op perfrorm method: #1115 and #1221

I would consider `stochastic_L_op` as the last resource, but I'm not against it.

**Reply by lucianopaz** [reply]
I like the rewrite approach, but how would that work with things like Binomial? Jax seems to implement the entire sampling method using jax primitives, but that's not the case for pytensor.

**Reply by ricardoV94** [reply]
In your use case do you need gradients wrt to likelihood as well? Discrete variables are usually not differentiable right?

**Reply by ricardoV94** [reply]
Fallback can an OpFromGraph with custom LOp which is basically the schocastic_L_op suggestion with indirection. If we find many such cases then it is a stronger argument towards `stochastic_L_op` method at the Op level

**Reply by lucianopaz** [reply]
My use case doesn't really have discrete variables, but since I'd like to have a general purpose tool I want to cover reparametrization tricks like Gumbel-max or Gumbel-softmax that enable stochastic gradients through some discrete variables.

**Reply by ricardoV94** [reply]
Right but that falls off from the rewrite approach still? I was surprised by the binomial as I'm not familiar with the trick for it if there's any, other than normal approximation for large n I guess

**Reply by jessegrabowski** [reply]
I would expect that discrete RVs don't have well-defined gradients, like all discrete functions
